src/toil/cwl/cwltoil.py

```diff
@@ -1805,6 +1806,77 @@ def determine_load_listing(tool: ToilCommandLineTool):
         )
     return load_listing
 
+class NoAvailableJobStoreException(Exception):
+    """Indicates that no job store name is available."""
+    pass
+
+def generate_default_job_store(batch_system_name: Optional[str], provisioner_name: Optional[str],
+                               local_directory: str) -> str:
+    """
+    Choose a default job store appropriate to the requested batch system and
+    porvisioner, and installed modules. Raises an error if no good default is
+    available and the user must choose manually.
+
+    :param batch_system_name: Registry name of the batch system the user has
+           requested, if any.
+    :param provisioner_name: Name of the provisioner the user has requested,
+           if any, May be 'aws' or 'gce'.
+    :param local_directory: Path to a nonexistent local directory suitable for
+           use as a file job store.
+
+    :return str: Job store specifier for a usable job store.
+    """
+
+    # Apply default batch system
+    batch_system_name = batch_system_name or DEFAULT_BATCH_SYSTEM
+
+    # Work out how to describe where we are
+    situation = f"the '{batch_system_name}' batch system"
+    if provisioner_name:
+        situation += f"with the '{provisioner_name}' provisioner"
+
+    try:
+        if provisioner_name == 'gce':
+            # We can't use a local directory on Google cloud
+
+            # Make sure we have the Google job store
+            from toil.jobStores.googleJobStore import GoogleJobStore
+
+            # Look for a project
+            project = os.getenv('TOIL_GOOGLE_PROJECTID')
+            project_part = (':' + project) if project else ''
+
+            # Roll a randomn bucket name, possibly in the project.
+            return f'google{project_part}:toil-cwl-{str(uuid.uuid4())}'
+        elif provisioner_name == 'aws' or batch_system_name in {'mesos', 'kubernetes'}:
+            # We can't use a local directory on AWS or on these cloud batch systems.
+            # If we aren't provisioning on Google, we should try an AWS batch system.
+
+            # Make sure we have AWS
+            from toil.jobStores.aws.jobStore import AWSJobStore
+
+            # Find a region
+            from toil.provisioners.aws import get_current_aws_region
+            region = get_current_aws_region()
+
+            if not region:
+                # We can't generate an AWS job store without a region
+                situation += ' running outside AWS with no TOIL_AWS_ZONE set'
+                raise NoAvailableJobStoreException()
+
+            # Roll a random name
+            return f'aws:{region}:toil-cwl-{str(uuid.uuid4())}'
```

Comment:
Should an `else` error with an unrecognized batch system if not `aws`/`gce`/`None`?

Reply:
I'll tell it to demand an explicit job store if it sees a provisioner that it doesn't know about.